src/coreclr/jit/assertionprop.cpp

```diff
@@ -3177,6 +3177,67 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* tree)
 
     if (conValTree != nullptr)
     {
+        if (tree->OperIs(GT_LCL_VAR) && conValTree->OperIs(GT_CNS_VEC, GT_CNS_DBL))
+        {
+            // A simple heuristic: If the constant is defined outside of a loop (not far from its head)
+            // and is used inside it - don't propagate.
+
+            bool keepPropagating = false;
+
+            if (conValTree->OperIs(GT_CNS_VEC))
+            {
+                if (conValTree->IsVectorZero() || conValTree->IsVectorAllBitsSet())
+                {
+                    // These are cheap to materialize
+                    keepPropagating = true;
+                }
+            }
+            else if (conValTree->OperIs(GT_CNS_DBL))
+            {
+                if (conValTree->IsFloatPositiveZero())
+                {
+                    // This is cheap to materialize
+                    keepPropagating = true;
+                }
+#if defined(TARGET_ARM64)
+                if (emitter::canEncodeFloatImm8(conValTree->AsDblCon()->gtDconVal))
+                {
+                    // Such floats are likely immable on arm64
+                    keepPropagating = true;
+                }
+#endif
+            }
+
+            if (!keepPropagating)
+            {
+                // Try to find the block this constant was originally defined in
+                const unsigned lclNum = tree->AsLclVarCommon()->GetLclNum();
+                const unsigned ssaNum = GetSsaNumForLocalVarDef(tree);
+                if (ssaNum != SsaConfig::RESERVED_SSA_NUM)
+                {
+                    BasicBlock* defBlock = lvaTable[lclNum].GetPerSsaData(ssaNum)->GetBlock();
```

Comment:
Style nit: `lvaTable[lclNum].` -> `lvaGetDesc(lclNum)->`.

General comment on the algorithm: this walks only one step up in the def chain, so if we have something like this:
```cs
var a = const;
for (...) {
    var b = a;
    Use(b);
}
```
We will still propagate `const` to `Use`.

Reply:
@SingleAccretion I don't see why it would.
When we're at Use(b) and we query SsaData()->Block from b we get `a`'s Block and it leads to "avoid propagating" path

Reply:
Ah, actually for your case `b` is going to be replaced with `a` in an early phase by `VN based copy prop`

Reply:
>  When we're at `Use(b)` and we query `SsaData()->Block` from `b` we get `a`'s Block and it leads to "avoid propagating" path

If I annotate this with blocks:
```cs
BB01:
var a = const;

for (...) {
BB02:
    var b = a; // b:d:1
    Use(b);    // b:u:1
}
```
I am pretty sure that when we query the def block for `b:u:1`, we'd get `BB02`, since the def `b:d:1` occurs inside it.

> Ah, actually for your case `b` is going to be replaced with `a` in an earlier phase VN based copy prop

I agree that copy propagation helps us here; not sure by how much because of the liveness constraints it has.

In any case, I also agree that there is no reason, for now, to complicate the code with a full UD chain walk, because we are pretty much targeting people hoisting constants out of the loop by hands.